### ISSUE TITLE
Any ideas about this?

### ISSUE THREAD

**Issue by nicpaimnt** [rossmacarthur/kb-remap#8]
I use kb-remap on both my work mac and my home mac. I swap between the two by 1 usb-c dock.

Keyboards, I have a few of them. Leopold, Filco etc. 

I share the same automator action to run a shell command such as this

`kb-remap --name "Mini Keyboard" --swap 0x700000039:0x700000068 --swap 0x700000049:0x70000004B --swap 0x70000004C:0x70000004E --swap 0x7000000E4:0x700000080 --swap 0x7000000E7:0x700000081`

What I don't understand is all things being the same, except the version of kb-remap, the home mac works (mini m2) but the work mac doesn't (air m3).

the output looks correct though. But the work mac has the latest kb-remap version 0.3.4. The home mac has an older version that doesn't have `--version`.

`Selected:
  Vendor ID: 0x853
  Product ID: 0x134
  Name: Mini Keyboard

Applied the following modifications:
  Raw(30064771129) -> Raw(30064771176)
  Raw(30064771176) -> Raw(30064771129)
  Raw(30064771145) -> Raw(30064771147)
  Raw(30064771147) -> Raw(30064771145)
  Raw(30064771148) -> Raw(30064771150)
  Raw(30064771150) -> Raw(30064771148)
  Raw(30064771300) -> Raw(30064771200)
  Raw(30064771200) -> Raw(30064771300)
  Raw(30064771303) -> Raw(30064771201)
  Raw(30064771201) -> Raw(30064771303)`

So I'm afraid of upgrading kb-remap on the home mac as it may not be working anymore. The output looks 'correct' at first glance. Any idea? Both run latest Sonoma...

to sum up, I find the new 0.3.4 doesn't work at all, even though it doesn't error out, on a macbook air m3.

thanks!

**Comment by rossmacarthur** [comment]
Here are some things to try, in the order I would try them

**Confirm that you are selecting the correct keyboard**

The keyboard names tend to be different between different Mac products. And sometimes what you think is the keyboard is not actually the keyboard (e.g. other devices appear as "keyboards"). You can list the keyboards using the following command:

```
kb-remap --list
```

If you are using the same `kb-remap` command in conjunction with the the `--name` option then the command that works on one machine might not work on another. The `--name "Mini Keyboard" ` option in your command likely won't work on your M3. 

You also don't have to specify `--name` and then the configuration will be applied to all keyboards.


**Confirm whether it is related to kb-remap version**

It would help to know the version of the older `kb-remap`. If you installed it with cargo you should be able to use the following command to list the installed crates and their versions. You could run the following on your Mac Mini to find the installed version.
```
cargo install --list
```

The `--version` option was added in `kb-remap v0.3.0`, so you could try installing the older version on your M3 and see if it works.
```
cargo install kb-remap --version '=0.2.1' --force
```

**Confirm your macOS versions**

Apparently the method that all versions of `kb-remap` uses to remap keys is broken on macOS 14.2, https://www.reddit.com/r/MacOS/comments/18g4vxn/cannot_remap_keys_on_macbook_pro_with_hidutils_in/. It appears it was fixed in macOS 14.3.

**Comment by nicpaimnt** [comment]
Thanks for the quick and thoughtful reply, much appreciated.

confirmed that `kb-remap --list` has "Mini Keyboard" (the Leopold FC660C).

So I'm on the work mac, and I have done the forced install of `0.2.1`, and the same command works now (volume control etc.). So it appears that it's a change in `0.3.0` that prevents it to work.

So I'm happy as Larry, as it works with this version.

That said, small correction, the work mac is an M2 **_as well as my home mac_**. Here are the details if that helps...

```system_profiler SPSoftwareDataType SPHardwareDataType
Software:

    System Software Overview:

      System Version: macOS 14.5 (23F79)
      Kernel Version: Darwin 23.5.0
      Boot Volume: Macintosh HD
      Boot Mode: Normal
      Computer Name: <redacted>
      User Name: <redacted>
      Secure Virtual Memory: Enabled
      System Integrity Protection: Enabled
      Time since boot: 1 day

Hardware:

    Hardware Overview:

      Model Name: MacBook Air
      Model Identifier: Mac14,15
      Model Number: Z18U0001XX/A
      Chip: Apple M2
      Total Number of Cores: 8 (4 performance and 4 efficiency)
      Memory: 16 GB
      System Firmware Version: 10151.121.1
      OS Loader Version: 10151.121.1
      Serial Number (system): <redacted>
      Hardware UUID: <redacted>
      Provisioning UDID: <redacted>
      Activation Lock Status: Disabled
```
      
Now that I know how to force install this version, I will upgrade my home mac later to `0.3.0` and see if it's an M2 problem or a difference between a _macbook air_ and a _mac mini_ since both are M2.

Thanks again, I will report later.

**Comment by rossmacarthur** [comment]
Okay I figured it out, using the raw key ids should not include the usage page as well, so basically this will work equivalently on both versions if you change the command to something like this:

 ```
 kb-remap --name "Mini Keyboard" --swap 0x39:0x68 --swap 0x49:0x4B --swap 0x4C:0x4E --swap 0xE4:0x80 --swap 0xE7:0x81
 ```
 
 I'm going to update the code to allow specifying the usage page and the usage id

**Comment by rossmacarthur** [comment]
Fixed in d79d9d93bc2ae173e4eb1d415e38040180124a41 and released in 0.4.0

**Comment by nicpaimnt** [comment]
Confirm that's a fix

```
$ cargo install kb-remap --version '=0.4.0' --force
... [snip]
Replaced package `kb-remap v0.2.1` with `kb-remap v0.4.0` (executable `kb-remap`)

$ kb-remap --version
kb-remap 0.4.0

$ kb-remap --reset
Reset all modifications

$ kb-remap --name "USB Keyboard" --swap 0x700000039:0x700000068 --swap 0x700000048:0x700000080 --swap 0x700000046:0x700000081
Selected:
  Vendor ID: 0x2efd
  Product ID: 0x2130
  Name: USB Keyboard

Applied the following modifications:
  Raw(30064771129) -> Raw(30064771176)
  Raw(30064771176) -> Raw(30064771129)
  Raw(30064771144) -> Raw(30064771200)
  Raw(30064771200) -> Raw(30064771144)
  Raw(30064771142) -> Raw(30064771201)
  Raw(30064771201) -> Raw(30064771142)  
```

Validated working (different keyboard).

Keys now function as expected in 0.4.0.

Thanks!
